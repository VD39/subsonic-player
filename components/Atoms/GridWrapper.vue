<script setup lang="ts">
withDefaults(
  defineProps<{
    desktop?: string;
    mobile?: string;
    tablet?: string;
  }>(),
  {
    desktop: '5',
    mobile: '2',
    tablet: '3',
  },
);
</script>

<template>
  <section
    :class="$style.mediaListWrapper"
    :style="{
      '--loop-rows-mobile': mobile,
      '--loop-rows-tablet': tablet,
      '--loop-rows-desktop': desktop,
    }"
  >
    <slot />
  </section>
</template>

<style module>
.mediaListWrapper {
  --loop-columns: var(--loop-rows-mobile);
  --loop-grid-gap: var(--space-24);

  position: relative;
  display: grid;
  grid-template-columns: repeat(var(--loop-columns), 1fr);
  gap: var(--loop-grid-gap);

  @media (--tablet-up) {
    --loop-columns: var(--loop-rows-tablet);
  }

  @media (width >= 1100px) {
    --loop-columns: var(--loop-rows-desktop);
  }
}
</style>
