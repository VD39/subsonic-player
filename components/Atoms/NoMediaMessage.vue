<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="['column', 'centerItems', $style.content]">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: var(--width-height-100);
}

.content {
  position: absolute;
  inset: 50% auto auto 50%;
  transform: translate(-50%, -50%);
}
</style>
