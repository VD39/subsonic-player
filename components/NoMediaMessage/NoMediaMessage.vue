<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="$style.content">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: 100%;
}

.content {
  @mixin align-center;

  position: absolute;
  inset: 50% auto auto 50%;
  flex-direction: column;
  transform: translate(-50%, -50%);
}
</style>
