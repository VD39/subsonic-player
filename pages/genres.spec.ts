import type { VueWrapper } from '@vue/test-utils';

import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { mount } from '@vue/test-utils';

import GridWrapper from '@/components/Atoms/GridWrapper.vue';
import NoMediaMessage from '@/components/Atoms/NoMediaMessage.vue';
import { getFormattedGenresMock } from '@/test/helpers';
import { useHeadMock } from '@/test/useHeadMock';

import GenresPage from './genres.vue';

const getGenresMock = vi.fn();

mockNuxtImport('useGenre', () => () => ({
  getGenres: getGenresMock,
}));

const genresDataMock = ref<{ genres: Genre[] }>({
  genres: [],
});

mockNuxtImport('useAsyncData', () => () => ({
  data: genresDataMock,
  status: ref('success'),
}));

const { useHeadTitleMock } = useHeadMock();

function factory(props = {}) {
  return mount(GenresPage, {
    global: {
      stubs: {
        GenreLink: true,
      },
    },
    props: {
      ...props,
    },
  });
}

describe('genres', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    genresDataMock.value = {
      genres: [],
    };
    wrapper = factory();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('sets the useHead function with correct title', () => {
    expect(useHeadTitleMock.value).toBe('Genres');
  });

  describe('when getGenres does not return any data', () => {
    it('does not show the GridWrapper component', () => {
      expect(wrapper.findComponent(GridWrapper).exists()).toBe(false);
    });

    it('shows the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(true);
    });
  });

  describe('when getGenres does return data', () => {
    beforeEach(() => {
      genresDataMock.value = {
        genres: getFormattedGenresMock(3),
      };

      wrapper = factory();
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the GridWrapper component', () => {
      expect(wrapper.findComponent(GridWrapper).exists()).toBe(true);
    });

    it('shows the correct number of genre items', () => {
      expect(
        wrapper
          .findComponent(GridWrapper)
          .findAll('[data-test-id="genre-item"]').length,
      ).toBe(3);
    });

    it('does not show the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(false);
    });
  });
});
