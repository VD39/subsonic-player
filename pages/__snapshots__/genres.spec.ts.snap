// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`genres > matches the snapshot 1`] = `
"<h1>Genres</h1>
<div class="column centerAll loadingData" style="display: none;"><span class="mainLoader"><span class="headphones"><svg fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.25" viewBox="0 0 24 24" width="75" xmlns="http://www.w3.org/2000/svg"><path d="M4 13.5V13c0-4.97 3.582-9 8-9s8 4.03 8 9v.5"></path><path d="M2 17.439v-1.877a2 2 0 0 1 1.515-1.94L4 13.5l1.254-.314a.6.6 0 0 1 .746.582v5.464a.6.6 0 0 1-.746.582l-1.74-.435A2 2 0 0 1 2 17.439ZM22 17.439v-1.877a2 2 0 0 0-1.515-1.94L20 13.5l-1.255-.314a.6.6 0 0 0-.745.582v5.464a.6.6 0 0 0 .745.582l1.74-.435A2 2 0 0 0 22 17.439Z"></path></svg></span><span class="centerItems loader"><span class="bar"></span><span class="bar"></span><span class="bar"></span><span class="bar"></span></span><span class="visuallyHidden">Loading data</span></span></div>
<div class="mainContent">
  <div class="column centerAll noMediaMessage">
    <phwaveform size="50"></phwaveform>
    <p>No genres found.</p>
  </div>
</div>"
`;

exports[`genres > when getGenres does return data > matches the snapshot 1`] = `
"<h1>Genres</h1>
<div class="column centerAll loadingData" style="display: none;"><span class="mainLoader"><span class="headphones"><svg fill="none" stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.25" viewBox="0 0 24 24" width="75" xmlns="http://www.w3.org/2000/svg"><path d="M4 13.5V13c0-4.97 3.582-9 8-9s8 4.03 8 9v.5"></path><path d="M2 17.439v-1.877a2 2 0 0 1 1.515-1.94L4 13.5l1.254-.314a.6.6 0 0 1 .746.582v5.464a.6.6 0 0 1-.746.582l-1.74-.435A2 2 0 0 1 2 17.439ZM22 17.439v-1.877a2 2 0 0 0-1.515-1.94L20 13.5l-1.255-.314a.6.6 0 0 0-.745.582v5.464a.6.6 0 0 0 .745.582l1.74-.435A2 2 0 0 0 22 17.439Z"></path></svg></span><span class="centerItems loader"><span class="bar"></span><span class="bar"></span><span class="bar"></span><span class="bar"></span></span><span class="visuallyHidden">Loading data</span></span></div>
<div class="mainContent">
  <section class="mediaListWrapper" style="--loop-rows-mobile: 2; --loop-rows-tablet: 3; --loop-rows-desktop: 5;">
    <genre-link-stub name="Rock" data-test-id="genre-item"></genre-link-stub>
    <genre-link-stub name="Rock" data-test-id="genre-item"></genre-link-stub>
    <genre-link-stub name="Rock" data-test-id="genre-item"></genre-link-stub>
  </section>
</div>"
`;
